### PR TITLE
awslogs driver: retry on all errors to PutLogEvents

### PR DESCRIPTION
**Pull request by sparrc**
**- What I did**

This commit makes a few changes:
1. rather than only retrying on an invalid sequence token, always retry
on PutLogEvents errors
2. fix double-logging of errors (the same errors were being logged in both
putLogEvents and publishBatch)
3. log the number of events being dropped if publishBatch fails
4. helper function for translating awserr.Error into logrus fields

**- How to verify it**

I manually instrumented the code to fail on PutLogEvents. Previous to this change logs look like this:

```
Oct 22 01:09:34 ip-10-0-0-253.us-west-2.compute.internal dockerd[23099]: time="2020-10-22T01:09:34.376056685Z" level=error msg="Failed to put log events" errorCode=EOF logGroupName=helloworld logStreamName=logs/main/XXX message="error putting logs" origError="original error foo"
Oct 22 01:09:34 ip-10-0-0-253.us-west-2.compute.internal dockerd[23099]: time="2020-10-22T01:09:34.376093409Z" level=error msg="EOF: error putting logs\ncaused by: original error foo"
```

after the change looks like this:

```
Oct 22 00:40:58 ip-10-0-0-253.us-west-2.compute.internal dockerd[17645]: time="2020-10-22T00:40:58.086032378Z" level=error msg="Failed to put log events, dropping batch of 5 events" errorCode=EOF logGroupName=helloworld logStreamName=logs/main/76bbf34ed2e447998951b15af41dd219 message="error putting logs" origError="original err foo"
```

**- Description for the changelog**

awslogs driver: retry all failed PutLogEvents requests, log number of dropped events in failed batches

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: daemon/logger/awslogs/cloudwatchlogs.go [hunk 1]
```diff
@@ -662,30 +650,22 @@ func (l *logStream) publishBatch(batch *eventBatch) {
...
 
 	nextSequenceToken, err := l.putLogEvents(cwEvents, l.sequenceToken)
-
 	if err != nil {
-		if awsErr, ok := err.(awserr.Error); ok {
-			if awsErr.Code() == dataAlreadyAcceptedCode {
-				// already submitted, just grab the correct sequence token
-				parts := strings.Split(awsErr.Message(), " ")
-				nextSequenceToken = &parts[len(parts)-1]
-				logrus.WithFields(logrus.Fields{
```
Comment (sparrc) [inline daemon/logger/awslogs/cloudwatchlogs.go:672]: this was previously not logging origError, so that is also fixed

**Comment by sparrc** [comment]
Hello, to give some extra context on this change, I am from the amazon ecs agent team (https://github.com/aws/amazon-ecs-agent) and we use this plugin extensively.

Sometimes our customers find that their logs are missing for a time, and all we can find in the logs are transient errors. We would like to add a retry in the PutLogEvent api call to hopefully hide any transient errors that might occur when writing a batch of events to cloudwatch.

When we do have to drop a batch of events, we'd like to also have visibility into how many events were dropped in order to have some idea about the number of missing logs we can expect to see in cloudwatch.

I considered adding more robust retrying logic and backoff, but that seems to me like a change that might have larger implications, because it would block the resetting of the buffer and the consumption of more messages from the `l.messages` channel. I went with this option because we already retry once in certain situations, so this change didn't seem too drastic a change in behavior.

**Comment by sparrc** [comment]
Jenkins is now only failing on TestInspect, which I dont think is related to my change:

```
[2020-10-22T16:13:47.373Z] === Failed
[2020-10-22T16:13:47.373Z] === FAIL: amd64.integration.service TestInspect (12.08s)
[2020-10-22T16:13:47.373Z]     inspect_test.go:39: timeout hit after 10s: waiting for tasks to enter run state. task failed with error: task: non-zero exit (1)
[2020-10-22T16:13:47.373Z] 
...

[2020-10-22T16:11:49.703Z] --- FAIL: TestInspect (12.08s)
[2020-10-22T16:11:49.703Z]     inspect_test.go:39: timeout hit after 10s: waiting for tasks to enter run state. task failed with error: task: non-zero exit (1)
```

**Review by samuelkarp — CHANGES_REQUESTED** [review]
Before I take a detailed look at the code, I chatted with @sparrc offline.  There are built-in retries in the AWS SDK for Go (iirc the default is 1 retry); if we need a different retry policy we should be overriding it in the SDK rather than wrapping the call to PutLogEvents.  The reason there is an existing wrapped retry is that the arguments need to change in the case of an invalid sequence token.

**Comment by djcoder100** [comment]
Is there an update on this PR and an original issue that this PR is related to?

**Comment by thaJeztah** [comment]
@samuelkarp @sparrc should we close this for now? (looking at your comment, it looks like it should be fixed in the SDK?)

**Pull request closed without merging**